src/backend/distributed/master/master_delete_protocol.c

```diff
@@ -163,7 +167,208 @@ master_apply_delete_command(PG_FUNCTION_ARGS)
+	bool isTopLevel = true;
+	DeleteStmt *deleteStatement = NULL;
+	bool failOK = false;
+	Oid relationId = InvalidOid;
+	List *queryTreeList = NIL;
+	Query *deleteQuery = NULL;
+	Node *whereClause = NULL;
+	Node *deleteCriteria = NULL;
+	LOCKTAG lockTag;
+	bool sessionLock = false;
+	bool dontWait = false;
+	List *shardIntervalList = NIL;
+	Index tableId = 1;
+	List *restrictClauseList = NIL;
+	List *prunedShardIntervalList = NIL;
+	ListCell *shardIntervalCell = NULL;
+	List *connectionList = NIL;
+	int32 affectedTupleCount = 0;
+	List *tableFailedPlacementList = NIL;
+	ListCell *tableFailedPlacementCell = NULL;
+	Node *queryTreeNode;
+
+	PreventTransactionChain(isTopLevel, "master_delete_from_shards");
+
+	queryTreeNode = ParseTreeNode(queryString);
+	if (!IsA(queryTreeNode, DeleteStmt))
+	{
+		ereport(ERROR, (errmsg("query \"%s\" is not a delete statement",
+							   queryString)));
+	}
+
+	deleteStatement = (DeleteStmt *) queryTreeNode;
+	relationId = RangeVarGetRelid(deleteStatement->relation, NoLock, failOK);
+
+	CheckDistributedTable(relationId);
+
+	queryTreeList = pg_analyze_and_rewrite(queryTreeNode, queryString, NULL, 0);
+	deleteQuery = (Query *) linitial(queryTreeList);
+	CheckTableCount(deleteQuery);
+
+	/* get where clause and flatten it */
+	whereClause = (Node *) deleteQuery->jointree->quals;
+	deleteCriteria = eval_const_expressions(NULL, whereClause);
+	CheckDeleteCriteria(deleteCriteria);
+
+	SET_LOCKTAG_ADVISORY(lockTag, MyDatabaseId, relationId, 0, 0);
+	LockAcquire(&lockTag, ExclusiveLock, sessionLock, dontWait);
+
+	shardIntervalList = LoadShardIntervalList(relationId);
+	restrictClauseList = WhereClauseList(deleteQuery->jointree);
+
+	prunedShardIntervalList =
+		PruneShardList(relationId, tableId, restrictClauseList, shardIntervalList);
+
+	PG_TRY();
+	{
+		foreach(shardIntervalCell, prunedShardIntervalList)
+		{
+			ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+			List *shardPlacementList = NIL;
+			ListCell *shardPlacementCell = NULL;
+			StringInfo shardQueryString = makeStringInfo();
+			Oid relationId = shardInterval->relationId;
+			uint64 shardId = shardInterval->shardId;
+			List *shardFailedPlacementList = NIL;
+			int32 shardAffectedTupleCount = -1;
+
+			deparse_shard_query(deleteQuery, relationId, shardId, shardQueryString);
+			shardPlacementList = FinalizedShardPlacementList(shardId);
+
+			foreach(shardPlacementCell, shardPlacementList)
+			{
+				ShardPlacement *shardPlacement =
+					(ShardPlacement *) lfirst(shardPlacementCell);
+				char *workerName = shardPlacement->nodeName;
+				uint32 workerPort = shardPlacement->nodePort;
+				char *nodeUser = CurrentUserName();
+				PGconn *connection = ConnectToNode(workerName, workerPort, nodeUser);
+				TransactionConnection *transactionConnection = NULL;
+				PGresult *result = NULL;
+				char *placementAffectedTupleString = NULL;
+				int32 placementAffectedTupleCount = -1;
+
+				if (connection == NULL)
+				{
```

Comment:
This function should be broken up. We try to keep all our code within 3 levels of nesting.